mammoth/modules/adapters.py

```diff
@@ -0,0 +1,250 @@
+        max_layer_index = max(adapter._adapted_layer_indices)
+        n_feedforwards = sum(layer_type == 'f' for layer_type in self._base_layer_types)
+        if max_layer_index >= n_feedforwards:
+            raise ValueError(
+                f'Invalid layer number {max_layer_index} in Adapter "{adapter.name}". '
+                f'There are ony {n_feedforwards} layers.'
+            )
+        self.adapters[adapter.name] = adapter
+
+    def deactivate_adapters(self):
+        self.active = set()
+
+    def activate_adapter(self, adapter_group: str, sub_id: str):
+        name = Adapter._name(adapter_group, sub_id)
+        if name in self.adapters:
+            self.active.add(name)
+
+    def _merge_active_adapters(self):
+        """
+        Returns a single mapping layer_idx -> list of AdapterLayer,
+        containing the layers of all currently active adapters
+        """
+        active_adapters = [
+            adapter for name, adapter in self.adapters.items()
+            if name in self.active
+        ]
+        merged = defaultdict(list)
+        for adapter in active_adapters:
+            for layer_idx, layers in adapter.get_layers():
+                merged[layer_idx].extend(layers)
+        return merged
+
+    def _inject_adapters(self):
+        adapted_layer_types = []
+        adapted_layers = nn.ModuleList()
+        adapted_layer_dropouts = []
+        i = 0
+        for layer_type, layer_struct, layer_dropout in zip(
+            self._base_layer_types,
+            self._base_layers,
+            self._base_layer_dropouts,
+        ):
+            adapter_layers_by_index = self._merge_active_adapters()
+            if layer_type == 'f':
+                # Adapters apply to feedforward layers
+                adapter_layers = adapter_layers_by_index[i]
+                tmp_layer_types = [layer_type]
+                tmp_layer_structs = [layer_struct]
+                tmp_layer_dropouts = [layer_dropout]
+                for adapter_layer in adapter_layers:
+                    tmp_layer_types, tmp_layer_structs, tmp_layer_dropouts = adapter_layer.apply(
+                        tmp_layer_types, tmp_layer_structs, tmp_layer_dropouts
+                    )
+                adapted_layer_types.extend(tmp_layer_types)
+                adapted_layers.extend(tmp_layer_structs)
+                adapted_layer_dropouts.extend(tmp_layer_dropouts)
+                i += 1
+            else:
+                # Attention layers are unmodified
+                adapted_layer_types.append(layer_type)
+                adapted_layers.append(layer_struct)
+                adapted_layer_dropouts.append(layer_dropout)
+        self.layer_types = adapted_layer_types
+        self.layers = adapted_layers
+        self.layer_dropouts = adapted_layer_dropouts
+
+    def forward(self, *args, **kwargs):
+        self._inject_adapters()
+        return super().forward(*args, **kwargs)
+
+
+# def layer_intermediates_shapes(cache):
```

Comment:
remove?

Reply:
Or maybe move into some utils subpackage together with the fancy qui_logging. I suspect that we're not yet done looking at the shapes of the intermediary state, and that is quite a gnarly structure.